modules/apps/wiki/wiki-service/src/main/java/com/liferay/wiki/model/impl/WikiPageImpl.java

```diff
@@ -96,6 +98,34 @@ public WikiPage fetchRedirectPage() {
 			getNodeId(), getRedirectTitle());
 	}
 
+	@Override
+	public List<FileEntry> getApprovedAttachmentsFileEntries()
+		throws PortalException {
+
+		if (!WorkflowDefinitionLinkLocalServiceUtil.hasWorkflowDefinitionLink(
+				getCompanyId(), getGroupId(), WikiPage.class.getName())) {
+
+			return getAttachmentsFileEntries(
+				QueryUtil.ALL_POS, QueryUtil.ALL_POS);
+		}
+
+		List<FileEntry> fileEntries = new ArrayList<>();
+
+		for (FileEntry fileEntry :
+				getAttachmentsFileEntries(
+					QueryUtil.ALL_POS, QueryUtil.ALL_POS)) {
+
+			int value = DateUtil.compareTo(
+				fileEntry.getModifiedDate(), getStatusDate());
+
+			if (value <= 0) {
+				fileEntries.add(fileEntry);
+			}
+		}
+
+		return fileEntries;
+	}
+
```

Comment:
Is there any way to avoid creating this new method? For example, would it be possible to modify the original `getAttachments*()` method so that, depending on the status of the current page, would return all files or only the approved ones? For what I've seen, that should work fine.

Reply:
Hey @adolfopa,
There are some instances in which we want to retrieve all attachments regardless of workflow status, like when [viewing the draft wiki page](https://github.com/liferay/liferay-portal/blob/master/modules/apps/wiki/wiki-web/src/main/resources/META-INF/resources/wiki/edit_page_view_attachments.jsp#L27) or [moving the wiki page to the trash](https://github.com/liferay/liferay-portal/blob/master/modules/apps/wiki/wiki-service/src/main/java/com/liferay/wiki/service/impl/WikiPageLocalServiceImpl.java#L1737-L1742).  Making the change you suggested could also affect those scenarios, leading to incorrect results.

Also, I think it would be confusing if the `getAttachmentsFileEntries()` method took workflow into consideration, but the superseding methods like `getAttachmentsFileEntries(int start, int end)` and `getAttachmentsFileEntries(int start, int end, OrderByComparator<FileEntry> orderByComparator)` don't.

Further, there are [getAttachmentsFileEntriesCount() methods](https://github.com/liferay/liferay-portal/blob/master/modules/apps/wiki/wiki-service/src/main/java/com/liferay/wiki/model/impl/WikiPageImpl.java#L145-L171) which do not take workflow into consideration.

Perhaps if we override [the getAttachmentsFileEntries method with the most arguments](https://github.com/liferay/liferay-portal/blob/master/modules/apps/wiki/wiki-service/src/main/java/com/liferay/wiki/model/impl/WikiPageImpl.java#L129-L143), adding a `checkWorkflow` boolean argument, we could have a slightly cleaner solution.  The same can be done for the `getAttachmentsFileEntriesCount()` method, or we could have these methods go through the `getAttachmentsFileEntries()` method instead

I guess I should step back to ask "why you are hesitant about my initial changes?".  Do you not want an API change?  Do you want this logic to be applied in more locations?  Or some other reason?

Thanks for looking into this PR!

Reply:
Maybe I'm mistaken, but when viewing/editing a draft wiki page, page.getStatus() == STATUS_DRAFT, and when approved status == STATUS_APPROVED. By checking this condition when computing the set of attachmetns, we should be able to decide whether we return the full set of files, or only the approved ones; as a plus, this would require no API changes. It would be simply a matter of replacing the old logic with your new one.

About the count method, we could extend the query so that it filters them by date, so it shouldn't be a problem.

About the reason why I'm reluctant to forward this as is, it is simply that I think it is not necessary to have two different methods that do roughly the same thing. Also, the new method doesn't make much sense for approved pages, but it will be there anyway. I think this makes the API a bit more confusing than necessary.

Reply:
Hey @adolfopa, I'll try to answer your concerns as structured as possible:

> Maybe I'm mistaken, but when viewing/editing a draft wiki page, page.getStatus() == STATUS_DRAFT, and when approved status == STATUS_APPROVED. By checking this condition when computing the set of attachmetns, we should be able to decide whether we return the full set of files, or only the approved ones; as a plus, this would require no API changes. It would be simply a matter of replacing the old logic with your new one.

You are not mistaken; we can check the status of the page.  When viewing a page, we may need to adjust which attachments are returned, based on workflow and page status.  When editing a page, I believe we should return all attachments, regardless of status'.  We should be able to remove both approved and pending attachments from the draft.  Here's the logic I'm proposing:

|| Latest Status | Viewing a Wiki Page  | Editing a Wiki Page |
| ------------- | ------------- | ------------- | ------------- |
| Workflow Enabled ||||
|| Published | Show all attachments | Show all attachments |
|| Not Published | Show all attachments added at or before last publish date | Show all attachments |
| Workflow Disabled ||||
|| Approved | Show all attachments  | Show all attachments  |
|| Draft | Show all attachments added at or before last publish date  | Show all attachments  |

This will require a bit of rework on my end, but I'd like to get the expected outcomes exact before resending a PR.

> About the count method, we could extend the query so that it filters them by date, so it shouldn't be a problem.

Works for me.

> About the reason why I'm reluctant to forward this as is, it is simply that I think it is not necessary to have two different methods that do roughly the same thing. Also, the new method doesn't make much sense for approved pages, but it will be there anyway. I think this makes the API a bit more confusing than necessary.

I believe we will need to add at least one new method, as we might need to override the page status (like when editing) and just retrieve all attachments.  I'll work on a cleaner solution which hopefully resolves all your concerns, unless there is something fundamentally wrong with my analysis here.  Thanks!